### PR TITLE
allow standalone dotted ops, parse .op as (. op)

### LINKED ISSUE #34156: Lower standalone dotted infix operators as anonymous functions
I could've sworn we already had an issue tracking this, but I couldn't find it.  In short, I'd love to have support for standalone `(.+)` and friends to lower to `(args...)->(+).(args...)`.  This would enable their use in higher order functions and generally make them more useful.

They currently parse as identifiers — and can be assigned to! — so this would be a breaking minor change.

### PR DESCRIPTION
**Pull request by simeonschaub**
A second attempt at #34156, should supersede #35706. This implements @JeffBezanson's suggestion in https://github.com/JuliaLang/julia/pull/35706#issuecomment-624272159, that we parse `.op` as `(. op)`, when it isn't functioning as an operator. We then lower this to `Base.BroadcastFunction(op)`, just as in #35706. Could we run PkgEval on this once tests pass to see whether this qualifies as a minor change?
closes #34156

### CODE REVIEW

File: base/broadcast.jl [hunk 1]
```diff
@@ -1264,4 +1264,36 @@ end
...
+Can be created by just passing an operator preceded by a dot to a higher-order function.
+
+# Examples
+```jldoctest
+julia> a = [[1 3; 2 4], [5 7; 6 8]];
+
+julia> b = [[9 11; 10 12], [13 15; 14 16]];
+
+julia> map(.*, a, b)
+2-element Array{Array{Int64,2},1}:
```
Comment (simeonschaub) [inline base/broadcast.jl:1282]:
```suggestion
2-element Vector{Matrix{Int64}}:
```

File: base/broadcast.jl [hunk 2]
```diff
@@ -1264,4 +1264,36 @@ end
...
+true
+```
+"""
+struct BroadcastFunction{F} <: Function
+    f::F
+end
+
+@inline (op::BroadcastFunction)(x...; kwargs...) = op.f.(x...; kwargs...)
+
+Base.show(io::IO, op::BroadcastFunction) = print(io, BroadcastFunction, '(', op.f, ')')
```
Comment (JeffBezanson) [inline base/broadcast.jl:1296]: This should use `show` on `op.f`.

**Comment by JeffBezanson** [comment]
It didn't occur to me that with this, `:.+` and `:(.+)` are different. A bit of a gotcha, but I'm not sure if we can do anything about it. Are there any dot operators that require the parentheses to be parsed after `:`? I think the answer is no, which is good since at least you can always just delete the parens if you need to.

File: base/broadcast.jl (hunk 2 above)
Comment (simeonschaub) [reply base/broadcast.jl:1296]: Yes, makes sense. I even thought about printing it as `(.f)` if `f` is an operator, but that might be a bit too cute.

**Comment by simeonschaub** [comment]
You certainly now this better than me, but I think currently every operator, with or without the dot, should be it's own token, so you should never actually need those parentheses. Even then, these could still be constructed by using the `Symbol` constructor directly.

**Comment by simeonschaub** [comment]
From my side, if you don't have any further comments, this should now be ready to do a PkgEval run.

**Comment by simeonschaub** [comment]
The test failures seem unrelated.

**Comment by simeonschaub** [comment]
Would it be possible to do a PkgEval run now, so this can be discussed in triage tomorrow?

**Comment by JeffBezanson** [comment]
:+1: from triage.

**Review by JeffBezanson — APPROVED** [review]

**Comment by JeffBezanson** [comment]
@nanosoldier `runtests(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/37583#issuecomment-702440226):
```
NanosoldierError: error when preparing/pushing to report repo: failed process: Process(`git pull -X ours`, ProcessExited(128)) [128]

```
Unfortunately, the logs could not be uploaded.
cc @maleadt

**Comment by maleadt** [comment]
PkgEval has migrated systems; let's try this again:

@nanosoldier `runtests(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/37583#issuecomment-703056086) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/ae243746d3208d6726c483c72de9d4e61c1bcbcc/pkgeval/by_hash/f78793f_vs_defde49/report.md). cc @maleadt

**Comment by simeonschaub** [comment]
The failures all look unrelated to this PR. 🥳

**Pull request merged**